src/Runner.Worker/ActionRunner.cs

```diff
@@ -89,6 +89,27 @@ public async Task RunAsync()
                 ExecutionContext.Warning($"`pre` execution is not supported for local action from '{repoAction.Path}'");
             }
 
+            // Handle Composite Local Actions
+            // Need to download and expand the tree of referenced actions
+            if (handlerData.ExecutionType == ActionExecutionType.Composite &&
+                handlerData is CompositeActionExecutionData compositeHandlerData &&
+                Stage == ActionRunStage.Main &&
+                Action.Reference is Pipelines.RepositoryPathReference localAction &&
+                string.Equals(localAction.RepositoryType, Pipelines.PipelineConstants.SelfAlias, StringComparison.OrdinalIgnoreCase))
+            {
+                var actionManager = HostContext.GetService<IActionManager>();
+                var prepareResult = await actionManager.PrepareActionsAsync(ExecutionContext, compositeHandlerData.Steps, ExecutionContext.Id);
+
+                // Reload the data once caches are setup
+                definition = taskManager.LoadAction(ExecutionContext, Action);
+                ArgUtil.NotNull(definition, nameof(definition));
+                handlerData = definition.Data?.Execution;
+                ArgUtil.NotNull(handlerData, nameof(handlerData));
+
+                // Save container setup steps so we can reference them later
+                (handlerData as CompositeActionExecutionData).ContainerSetupSteps = prepareResult.ContainerSetupSteps;
```

Comment:
might want to make sure `prepareResult.ContainerSetupSteps` will never be null

Reply:
consider: is setting ContainerSetupSteps on the handlerData an abuse of that object? iirc that info is purely from the action.yml

consider passing the just-in-time container setup steps to the handler factory instead